sips/sip-018/sip-018-signed-structured-data.md

```diff
@@ -0,0 +1,204 @@
+# Preamble
+
+SIP Number: 018
+
+Title: Signed Structured Data
+
+Author: Marvin Janssen <https://github.com/MarvinJanssen>
+
+Consideration: Technical
+
+Type: Standard
+
+Status: Accepted
+
+Created: 28 December 2021
+
+License: CC0-1.0
+
+Sign-off: Jude Nelson <jude@stacks.org>, Technical Steering Committee Chair
+
+Layer: Applications
+
+Discussions-To: https://github.com/stacksgov/sips
+
+# Abstract
+
+The Signed Structured Data specification describes a standard way to deterministically encode and sign Structured Data. Structured Data is data that can be represented in human-readable format and used in applications and smart contracts. The aim of the standard is to produce signatures that are straightforward and inexpensive to verify by smart contracts.
+
+# License and Copyright
+
+This SIP is made available under the terms of the Creative Commons CC0 1.0 Universal license, available at https://creativecommons.org/publicdomain/zero/1.0/
+This SIP's copyright is held by the Stacks Open Internet Foundation.
+
+
+# Introduction
+
+Digital signatures are at the heart of blockchains. They allow users to trustlessly transfer assets, invoke smart contracts and more. To perform these actions, a user signs a transaction and broadcasts it to the network. However, there are situations in which signed proofs without having to broadcast transactions are desirable. A few common use-cases include: (1) prove to an external application or entity that a user is in control of an address; (2) authorise an action to be performed by a smart contract at a later stage (like a meta transaction); (3) participate in an off-chain mechanism that is later settled on-chain (like a subnet).
+
+It is important that signed messages are understandable for humans. For transactions, this is obvious: wallet applications display whom is receiving how many tokens of what kind. Likewise, the input parameters, function name, and target contract of contract calls are properly listed. Signed messages that are not transactions should be no different.
```

Comment:
I'm confused by the meaning of this paragraph.  Why is it undesirable that it's possible to post signed encrypted data (i.e. not understandable to humans) to a blockchain?

I think you might be trying to advocate for representing structured data as a Clarity value (even when alternative structured data representations exist, such as JSON), but this isn't making that case.

Reply:
I never stated it is undesirable that it is possible to post signed encrypted data to a blockchain. I stated that it is desirable that the user knows what he or she is signing when an app request a signature. Example: there is a lot of precedent in the Ethereum world where users were prompted to sign hashes of 0xProtocol order data and messing it up / getting scammed and losing their tokens. No way to verify that the order struct hash you are signing in your wallet app matches what you see on the app.

Here is what that looked like:
![bad](https://raw.githubusercontent.com/ethereum/EIPs/master/assets/eip-712/eth_sign.png) 

We don't want to go down that route.

Reply:
> I never stated it is undesirable that it is possible to post signed encrypted data to a blockchain.  I stated that it is desirable that the user knows what he or she is signing when an app request a signature

I don't know how to reconcile "It is important that signed messages are **understandable for humans**" (emphasis mine) with the above, in the context of describing a data representation that lives in the Clarity smart contract.  As written, the text is calling for a structured data schema that pertains to how _Clarity values_ are represented within the execution context of a _smart contract_.  Making sure that users know what they're signing is all well and good, but that's not what the text says.  In general, this is not something the Clarity VM can enforce, nor is it something that a structured data schema can ever hope to mandate (since "understandable to humans" is in the eye of the beholder, not the specification).

Can you change this text to clarify this?

Reply:
I think I understand what you're trying to say.  You're saying that in the context of users signing data with a key that potentially owns data on a blockchain, there are two important usability considerations:

* The user must be able to understand what it is they are signing, as well as the implications of publishing the signed message in a public space (such as a blockchain).

* Any software that supports signing data besides transactions should structure this data such that it is all but impossible for the resulting signature to be a signature for a valid transaction (this is usually achieved by signing both the data and a preamble, like what you propose here).

Part of this SIP is to propose a schema for point (2).  I think this is a worthy goal, but I also think it should be a separate SIP from the SIP that describes `to-consensus-buff`.  In general, we'd like to keep all the Stacks 2.1 SIPs as "small" as possible, so that they only describe the breaking changes they create.